### ISSUE TITLE
Add functionality to BoxAssembly checklist so InventoryAdjustments get created

### ISSUE THREAD

**Issue by BtheGit** [rubyforgood/voices-of-consent#20]
(See details below)

**Comment by maebeale** [comment]
Box Assembly (completing the checklist of associated BoxItems to confirm they were packed in the Box) should adjust the InventoryTally for the location of the AssembledBy volunteer.

For example:
- Mark "Toothbrush" as included in the Box on BoxAssembly page
- Determine the volunteer's Assembly Location (will need to add assembly_location_id foreign_key to the database)
- Controller creates an InventoryAdjustment, subtracting one toothbrush from the Assembly Location's InventoryTally

**Comment by maebeale** [comment]
Functionality desired:
- After clicking "check" on a checkbox, controller creates an InventoryAdjustment, subtracting one from the Assembly Location's InventoryTally for that InventoryType

Background:
- BoxAssembly is a checklist of BoxItems for the Assembler to physically put in the Box
- Checking an item on the Box Assembly page _should_ adjust the InventoryTally (for the location of the Assembler).

In more detail:
- Assembler marks the "Toothbrush" to indicate they've put it in the Box
- We need to add code to the BoxAssembly controller to:
  - Determine the Assembler's Location (select the first Location of type `StorageUnit` for now)
  - Find the InventoryTally that's connected to the Location you found, and the InventoryType checked ("Toothbrush")
  - Create an InventoryAdjustment that belongs to this InventoryTally in the quantity of -1

**Comment by oconnelc** [comment]
I can start taking a look at this. I'm setting up my workspace now and am working at a "Code For Good" meetup. I'm not sure how much time I can dedicate after today, but I want to see how far I get tonight.

**Comment by maebeale** [comment]
Great, @oconnelc !!!

**Comment by oconnelc** [comment]
I'm still looking at this. I was working on it today but was having problems setting up my environment. I think I've resolved the environment issues.

**Comment by maebeale** [comment]
No rush, @oconnelc ! Let us know if we can be of any help!

**Comment by oconnelc** [comment]
I'm trying to get my dev setup and have performed the following steps:

# Installed the following software
- ruby 2.6
- node 10+
- yarn 1.16+
- bundler 2+
# Ran the following commands
Running docker desktop by running the following command:
`docker-compose up -d`
and verified that it running by running the command:
`docker-compose ps`

Which gives me the output: 
![image](https://user-images.githubusercontent.com/1112973/70380284-6c8e8880-1906-11ea-8900-e3ab493b74bd.png)

# Docker Desktop IP Address 
Now because I'm on Windows and running the docker desktop, the docker sessions are running on a different IP. (example 192.XXX.XX.XXX). Usually instead of running localhost or 172.0.0.1, I use the IP that Docker Desktop provides. 

# Started trying to Run the App

- `bundle install`
- `yarn install`

## The Following Command Fails
`rake dev:setup`

With the following error:
$ rake dev:setup
```
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
could not connect to server: Connection refused (0x0000274D/10061)
        Is the server running on host "localhost" (::1) and accepting
        TCP/IP connections on port 5432?
could not connect to server: Connection refused (0x0000274D/10061)
        Is the server running on host "localhost" (127.0.0.1) and accepting
        TCP/IP connections on port 5432?
Couldn't create 'voices-of-consent_development' database. Please check your configuration.
rake aborted!
PG::ConnectionBad: could not connect to server: Connection refused (0x0000274D/10061)
        Is the server running on host "localhost" (::1) and accepting
        TCP/IP connections on port 5432?
could not connect to server: Connection refused (0x0000274D/10061)
        Is the server running on host "localhost" (127.0.0.1) and accepting
        TCP/IP connections on port 5432?

Tasks: TOP => dev:setup => db:setup => db:schema:load_if_ruby => db:create
```

I think the problem is that the DB is running on the fake machine that the Docker Desktop launches for the containers but something in the dev configuration is pointing to localhost. I wasn't able to find where this configuration lies. Can you help me get passed this problem?

**Comment by oconnelc** [comment]
Ok I got a little bit further. I updated the docker-compose.yml so that instead of localhost it's just 5432:5432 for the ports. I also updated the database.yml file in config to point to the docker desktop IP and it looks like it is connecting. 

However it's now failing with this error:
`rake dev:setup`

```
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
FATAL:  role "username" does not exist
Couldn't create 'voices-of-consent_development' database. Please check your configuration.
rake aborted!
ActiveRecord::NoDatabaseError: FATAL:  role "username" does not exist


Caused by:
PG::ConnectionBad: FATAL:  role "username" does not exist

Tasks: TOP => dev:setup => db:setup => db:schema:load_if_ruby => db:create
```

Where username is the name on the machine I'm using.

**Comment by oconnelc** [comment]
The lesson here is to make sure you read all warnings and errors. 

When I ran the docker compose to build the containers, including the database, I was getting the following error:

`WARNING: The USER variable is not set. Defaulting to a blank string.`

The docker desktop was not setting the USER environment variable, so the role was not being created in the DB. I ended up manually setting it to the correct value and rake was able to build all of the tables.

**Comment by oconnelc** [comment]
I've gotten up to the heroku command and I'm getting the following error:

```
$ heroku local -f Procfile.dev
2:53:02 PM web.1     |  'bin' is not recognized as an internal or external command,
2:53:02 PM web.1     |  operable program or batch file.
[DONE] Killing all processes with signal  SIGINT
2:53:02 PM webpack.1 |  'bin' is not recognized as an internal or external command,
2:53:02 PM webpack.1 |  operable program or batch file.
2:53:02 PM webpack.1 Exited with exit code null
2:53:02 PM web.1     Exited with exit code null
2:53:02 PM sidekiq.1 Exited with exit code SIGINT
```

I'm confused as to what could be causing this problem. If I run the commands with the Procfile.dev individually, they run without any problem. It seems like it's having problems understanding the command in the Procfile.

Can anyone help me get this to run?

**Comment by maebeale** [comment]
@oconnelc would you add the USER note as a PR update to the README? There's a gotchas section there. Ideally we'll be adding this Procfile situation to it before too long!

**Comment by maebeale** [comment]
Working on finding some help, @oconnelc !

**Comment by h-m-m** [comment]
@oconnelc For your latest comment, I wonder if a script is trying to run "bin/webpack" and something is interpreting the way the MS command prompt would as "bin /webpack" instead of the path to the executable, "bin\webpack"

I've not used the heroku commands on Windows yet. Is there a way to tell them the right way to interpret the Procfile? Does it work if you edit the Procfile commands?

**Comment by maebeale** [comment]
@oconnelc hello! were you able to make it past the above blocks?

**Comment by oconnelc** [comment]
Sorry for the delayed response. I got pulled onto something else and got a new laptop. I'm trying to set it up now. 

One other thing is that the README says to use Ruby 2.6.0 but when I try that I get an error saying that 2.6.3 is the version to use.

**Comment by oconnelc** [comment]
Ok now I'm stuck at the initial setup for the App. I've installed Ruby 2.6.5 from https://rubyinstaller.org/downloads/ and that installs a bundler at version 1.17.2. I run:

`gem install bundle`

and get the following output:
```
Fetching bundler-2.1.4.gem
Successfully installed bundler-2.1.4
Parsing documentation for bundler-2.1.4
Installing ri documentation for bundler-2.1.4
```

But when I try to run `bundle install` I get the following output:
```
Warning: the running version of Bundler (1.17.2) is older than the version that created the lockfile (1.17.3). We suggest you upgrade to the latest version of Bundler by running `gem install bundler`.
Your Ruby version is 2.6.5, but your Gemfile specified 2.6.3
```
Does anyone know how to resolve this?

**Comment by h-m-m** [comment]
Yes. Feel free to edit the Gemfile to say `ruby '2.6.5'`. The version difference is small enough that it should make no difference for this project. Enough other contributors this week have had trouble with this error that I think the version bump will save more headaches than it has caused

**Comment by oconnelc** [comment]
Ok so here's where I'm at:

1. I was able to get the DB and related docker containers deployed and running, after setting the environment variable $USER to match the role that's used in the code. It also looks like the code that's responsible for connecting to the DB is correctly pulling the username, so it looks like there could be a potential of a mismatch. 
2. Per @h-m-m 's recommendation, I updated the version of Ruby and it worked. 
3. I'm still having the Heroku problem with the Procfile.dev file. It looks like Heroku is not pointing to the bin directory correctly. I wonder if it's running in a different directory.

Long story short, don't code in Windows. I'm working on some documentation to cover all of the subtle things that have to be done to get this to work in Windows. 

I'm assuming I could just run the commands by hand; I'm going to try that next.

**Comment by oconnelc** [comment]
Running by hand seems to have started everything. For the rails server, I see:

```
$ bin/rails server
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
  Please add the following to your Gemfile to avoid polling for changes:
    gem 'wdm', '>= 0.1.0' if Gem.win_platform?
=> Booting Puma
=> Rails 5.2.3 application starting in development
=> Run `rails server -h` for more startup options
*** SIGUSR2 not implemented, signal based restart unavailable!
*** SIGUSR1 not implemented, signal based restart unavailable!
*** SIGHUP not implemented, signal based logs reopening unavailable!
Puma starting in single mode...
* Version 3.12.1 (ruby 2.6.5-p114), codename: Llamas in Pajamas
* Min threads: 5, max threads: 5
* Environment: development
* Listening on tcp://localhost:3000
Use Ctrl-C to stop
```
For the webpack, I'm seeing:
```
$ bin/webpack-dev-server --watch
i ｢wds｣: Project is running at http://localhost:3035/
i ｢wds｣: webpack output is served from /packs/
i ｢wds｣: Content not from webpack is served from C:\Users\cocon\workspace\voices-of-consent\public\packs
i ｢wds｣: 404s will fallback to /index.html
Browserslist: caniuse-lite is outdated. Please run next command `yarn upgrade`
i ｢wdm｣: Hash: ae6959cb18ffebbdfa94
Version: webpack 4.36.1
Time: 9988ms
Built at: 02/15/2020 4:02:52 PM
                                                    Asset      Size                      Chunks             Chunk Names
                   js/application-2a389ea12cbee21c0c03.js  2.74 MiB                 application  [emitted]  application
               js/application-2a389ea12cbee21c0c03.js.map  2.93 MiB                 application  [emitted]  application
             js/box_assembly_form-3615b7b7bc71b8d90410.js  1.35 MiB           box_assembly_form  [emitted]  box_assembly_form
         js/box_assembly_form-3615b7b7bc71b8d90410.js.map  1.54 MiB           box_assembly_form  [emitted]  box_assembly_form
               js/box_design_form-d1baa37a055da319872b.js  1.73 MiB             box_design_form  [emitted]  box_design_form
           js/box_design_form-d1baa37a055da319872b.js.map  1.89 MiB             box_design_form  [emitted]  box_design_form
              js/box_request_form-235b8beff3b78b7f1a26.js  1.46 MiB            box_request_form  [emitted]  box_request_form
          js/box_request_form-235b8beff3b78b7f1a26.js.map  1.63 MiB            box_request_form  [emitted]  box_request_form
                         js/embed-d4f8b8a1649950c089a8.js   434 KiB                       embed  [emitted]  embed
                     js/embed-d4f8b8a1649950c089a8.js.map   488 KiB                       embed  [emitted]  embed
                         js/login-b033a8efbf60e4124167.js  1.34 MiB                       login  [emitted]  login
                     js/login-b033a8efbf60e4124167.js.map  1.53 MiB                       login  [emitted]  login
              js/message_log_form-ad36a19622de3eede95e.js  1.38 MiB            message_log_form  [emitted]  message_log_form
          js/message_log_form-ad36a19622de3eede95e.js.map  1.57 MiB            message_log_form  [emitted]  message_log_form
    js/volunteer_application_form-cbd271f7ce8a2593353b.js  1.77 MiB  volunteer_application_form  [emitted]  volunteer_application_form
js/volunteer_application_form-cbd271f7ce8a2593353b.js.map  1.91 MiB  volunteer_application_form  [emitted]  volunteer_application_form
                                            manifest.json  3.29 KiB                              [emitted]
               media/webfonts/fa-regular-400-65b286af.ttf  33.3 KiB                              [emitted]
               media/webfonts/fa-regular-400-7b9568e6.svg   141 KiB                              [emitted]
               media/webfonts/fa-regular-400-c1a866ec.eot  33.6 KiB                              [emitted]
              media/webfonts/fa-regular-400-c4f508e7.woff  16.4 KiB                              [emitted]
             media/webfonts/fa-regular-400-f5f2566b.woff2  13.3 KiB                              [emitted]
i ｢wdm｣: Compiled successfully.
```

When I go to http://localhost:3035, I see the following message in the browser:
`Cannot GET /`

Is there an index? Am I in a broken state? Are there other views?

**Comment by h-m-m** [comment]
You're connecting to the webpack developer server (wds), not the Rails server. Several other people were also confused by this on Thursday.

If you scroll a little further back, you should see a line that indicates the web service which runs Rails is listening on port 5000 instead of 3035

**Comment by oconnelc** [comment]
OK sorry for not getting to this sooner and for having some many problems (on my end) with the setup. I was able to get the application to run and I was able to login with the debug account. 

So now I'm having some weird problems. Clicking on the following buttons at the top caused the following pages to load:

- Status -> http://localhost:3000/box_requests
- Review -> http://localhost:3000/
- Design -> http://localhost:3000/box_requests
- Packing -> http://localhost:3000/
- Shipping ->http://localhost:3000/

Is this expected behavior? 

Also for this ticket which page is this displayed?

**Comment by h-m-m** [comment]
Thank you for the work you've done with our setup process. It definitely can use some care and attention. The problems you had setting it up has helped us shine some light on where it's weakest, so I appreciate that.

The behavior you're seeing is a known issue. It should be fixed by https://github.com/rubyforgood/voices-of-consent/pull/414

**Comment by oconnelc** [comment]
I pulled down the fix for #414 and was getting the following error when filtering by research_in_progress and researched

http://localhost:3000/box_requests?utf8=%E2%9C%93&filter_by=research_in_progress
http://localhost:3000/box_requests?utf8=%E2%9C%93&filter_by=researched

![image](https://user-images.githubusercontent.com/1112973/74992846-46fa2100-5417-11ea-8da8-64a7b78bea5c.png)

**Comment by maebeale** [comment]
Hey there, @oconnelc ! You definitely get the marathon award! Way to stay with it!!! And thank you! #414 was part of it, but there were a couple other issues that is in the process of fixing this as well. We had to had the scopes to box request so they could be filtered by subsequent box status.

**Comment by oconnelc** [comment]
How can I see this functionality on the app? Is it under one of the tabs? If so, which one? 

What is a good way to populate the local DB so that I can test this capability?

**Comment by oconnelc** [comment]
Disregard my question about populating the DB. Also I think I found the BoxAssembly stuff, so I can start working on it.

**Comment by oconnelc** [comment]
I have some questions about the requirements for this issue:

1. ```Assembler marks the "Toothbrush" to indicate they've put it in the Box``` What does it mean to mark toothbrush? On my screen I see: 
![image](https://user-images.githubusercontent.com/1112973/75099793-c280dd00-5593-11ea-8503-5bc5fdbb8b0d.png)
Is marking changing the item count?

1. ```Determine the Assembler's Location (select the first Location of type StorageUnit for now)``` Where is the Assembler's location stored? I looked in the user table and didn't see any

**Comment by maebeale** [comment]
1. Hello! I think that pic indicates box design page. There should show a checkbox?  Are you at the /box_assembly endpoint?

2. You're right! It's not on user, though adding a user-location join table later will prob be a good idea. `assembly_location_id` should be an FK on box? Would be cool to have a dropdown on the page that sends an ajax update to the box so when the items are checked they can be deducted from the correct location. 

More questions welcomed! :)

**Comment by oconnelc** [comment]
@maebeale the problem I'm running into is that I can't seem to get a box into the "Box Assembly" state. For example with this box; http://localhost:3000/box_design/40 I get the following output:

![image](https://user-images.githubusercontent.com/1112973/75615769-6a5b5500-5b16-11ea-8e3d-c6fee7d827c5.png)
Clicking on the "The box is ready to go" does nothing. 

However I was able to get to the BoxAssembly view by going direct to: 
http://localhost:3000/box_assembly/40 so I think I'm able to proceed.

**Comment by maebeale** [comment]
@oconnelc hello! correct, the button doesn't work yet. i'm working on that issue currently. 
to get past box_design, go to this endpoint: http://localhost:3000/box_assembly/40/complete, and then you'll be at the assembly step, where what we want is for when you click the checkbox, it'll create an InventoryAdjustment to deprecate the quantities of that inventory_item. (you'll need to make an assumption about which storage_location to subtract from to start.)

you're almost there!!! thanks for hanging in there and making it through the thorns!!!

**Comment by oconnelc** [comment]
Ok I think I'm finally starting to understand this. 

To create an InventoryAdjustment, I should call on the InventoryAdjustment endpoint new, to create a new one, and then make the update on client side and push an update. 

The following still confuses me:

1. How do the methods in the controller classes get identified as GET or POST? 

1. How does the InventoryAdjustment update the inventory? I didn't see a trigger in the database. 

1. It looks like there is an endpoint that will create and update the InventoryAdjustment: create. It looks like it takes the parameters:

```
def inventory_adjustment_params
      params.require(:inventory_adjustment).permit(:inventory_tally_id,
                                                   :purchase_id,
                                                   :box_item_id,
                                                   :total_cost,
                                                   :adjustment_quantity)
    end
```
Does this mean that when an item is checked, the POST param should be: 

```
{
   box_item_id: <id from checkbox>,
   adjustment_quantity: -1
}
```

or

```
{
   inventory_adjustment: {
      box_item_id: <id from checkbox> ,
      adjustment_quantity: -1
   }
}
```

**Comment by maebeale** [comment]
Hello there, @oconnelc . Yup! Exactly. It should be an adjustment_quantity of -1.

1. These are Rails conventions as to get/post. If you visit `/rails/info/routes` it'll show that part
2.  The InventoryAdjustment *is* the db update. So, the `inventory_tally` is the original declaration of: "I've got 5 toothbrushes is storage_location #5." And subsequent InventoryAdjustments will change that number, so to find the current item total (at that location) we'd be calling :sum on the quantities of the original tally + all adjustments. It's essentially like a row in a ledger. (At one point we were going to cache the start and end totals so you don't need to map over all the db records to get the current totals, but we decided to start simply and only optimize if performance became a problem.)
3. Pretty sure the second one. It'll also need to send an `:inventory_tally_id` value so the adjustment knows which tally (storage_location + item_type) to be adjusting.

**Comment by maebeale** [comment]
@oconnelc there's room for your suggestion re whether these adjustments should *only* be created when the button at the bottom is clicked, or upon checking the checkbox. My vote is the checkbox, but there would need to be an opposite adjustment created when UNchecking the box... This can be a separate issue, or folded in to this one -- whatever you recommend!

**Comment by oconnelc** [comment]
@maebeale thank you for the response. That makes sense. I was able to pull down a location that is a storage_unit. I'm trying to figure out how create the InventoryAdjustment. I think I need to call the create endpoint with the following parameters:

- box_item_id: equal to the item.id
- adjustment_quantity: -1 or 1, depending on the checkbox state 

How is the location used? Am I supposed to create a new InventoryTally? If so, it looks like I would need to create a Inventory Type. If that's true what does that type mean? 

To summarize it looks like I need to create the following types:

# InventoryType
- name: not sure what this is
- description: not sure what this is
- requires_research: not sure what this is

# InventoryTally

- additional_location_info: not sure what this is
- inventory_type_id: id for the Inventory Type that was created
- storage_location_id: The ID of the storage (for now just using the first Storage Unit). 

# InventoryAdjustment
- inventory_tally_id: The id of the Inventory Tally created for this action
- purchase_id: not sure what this is
- box_item_id: The id of the item, found from the items checkboxes. 
- total_cost: not sure what this is
- adjustment_quantity: 1 or -1 based on the checkbox state

**Comment by maebeale** [comment]
@oconnelc hey there!

re `box_item_id: equal to the item.id`: BoxItem is its own model, so, you should have a Box with several BoxItems (that's what's being checked off). A BoxItem has a box_id and an inventory_type_id (and a quantity), connecting the InventoryType (e.g. toothbrush, to a specific Box).

In production, the data would already have InventoryTallies for each InventoryType of all the Box's BoxItems, but, yeah, we should create one if one doesn't exist (we'll need to clean up the storage_location_id situation later -- that needs to be indicated by the user at some point).

Have filled in info below (would you be willing to update `https://github.com/rubyforgood/voices-of-consent/blob/develop/notes/background.md` w whatever of the following info you think would help someone else walk in to this inventory segment of the code?)

 
# InventoryType
- name: [[things that get put in boxes, e.g. toothbrush, teddy bear, nail file]]]
- description: [[[some inventory_types, like a specific book, might have a summary of the book here]]]
- requires_research: [[[some types, like "legal info for your area," require research before they can be included in the Box. toothbrushes are false for this. :)]]]

# InventoryTally
- additional_location_info: [[[this is to indicate _where_ in the storage_location something might be located, like "Shelf #A2" or something]]]
- inventory_type_id: id for the Inventory Type that was created [[[we aren't creating types here -- those should be in seed data, but we are referencing existing types. think of this part as if you're just in charge of logging what you have in your specific storage unit: 5 toothbrushes, 3 welcome back books, 14 teddy bears, etc]]]
- storage_location_id: The ID of the storage (for now just using the first Storage Unit). [[[yup! first storage unit for now is great!]]]

# InventoryAdjustment
- inventory_tally_id: The id of the Inventory Tally created for this action [[[the inventory tally that's references for this action -- created only if necessary. ideally all the inventory tallies will exist to start. i'll create an issue so when creating a storage_location, you automatically get a tally for every existing inventory_type at a quantity of 0.]]]
- purchase_id: [[[this isn't relevant to BoxAssembly, but will be referenced when an InventoryAdjustment is a + amount (vs a - amount like here) bc someone went to the store and bought 10 new toothbrushes.]]]
- box_item_id: The id of the item, found from the items checkboxes. [[[yup!]]] 
- total_cost: [[[this is for accounting purposes and not relevant to the box_assembly]]]
- adjustment_quantity: 1 or -1 based on the checkbox state [[[yup!]]]

REALLY appreciate your attention to detail on this! VERY helpful questions/clarifications. Thanks, @oconnelc !

**Comment by oconnelc** [comment]
Sorry it's been so long since my last update. I'm hoping I should have some time this weekend and into this week to continue working. 

The ticket started with selecting the first storage unit location and show it on the top of the screen. I have that working. However it appears that there is only InventoryTally in my DB and that's for a university. Should I create the Tally if it doesn't exist for the location? 

Assuming that I have the tally, is the intent to create an InventoryAdjustment for the item and checkbox state?

**Comment by maebeale** [comment]
@oconnelc no worries!!! A lot of us are a little longer in our cycles than usual. :)  

Hmmm. Technically I think they should get an error saying they'll need to "restock" or something. 

But maybe for now create a tally at quantity 0 and the adjustment should go to -1 ??? 

Or maybe we just make seeds for a storage location now, and make another issue re prompting the user to review storage location inventory? 

Open to suggestions!

**Comment by oconnelc** [comment]
I went ahead and made the change that it will go out and try to get an Inventory Tally. Right now if it doesn't find one for the location, the check box won't change state. Let me know if there's anything else you'd like me to do for this ticket.

**Comment by maebeale** [comment]
Perfect! Yeah, I think this is great @oconnelc .

And when the Assembler clicks a button at the bottom "Complete" (or something like that), that's when it'll create the InventoryAdjustments against the InventoryTallies it found?

**Comment by oconnelc** [comment]
Actually the way I have it now is thae InventoryAdjustments get created whenever the checkboxes are clicked. I can submit an update tomorrow to add the Complete button and it's functionality.

**Comment by maebeale** [comment]
I realized that after looking at it in more depth.

Probably what you did is better than an official full POST through a button because Assembler will know in real time if there aren't any more available.

But the button to complete it can trigger the state machine to bring the Box to the next phase in the lifecycle, so we would def still want a button.
